ncwriter/schema.py

```diff
@@ -18,28 +17,53 @@
 # Create a new validator class (based on Draft4Validator) to allow templates to use
 # * Python types or numpy dtypes to specify variable data types; and
 # * numpy arrays to specify variable data.
-meta = deepcopy(Draft4Validator.META_SCHEMA)
-meta['definitions']['simpleTypes']['enum'].append('datatype')
-types = deepcopy(Draft4Validator.DEFAULT_TYPES)
-types['datatype'] = (basestring, type, np.dtype)
-types['array'] = (list, np.ndarray)
-TemplateValidator = validators.create(meta_schema=meta,
-                                      validators=Draft4Validator.VALIDATORS,
-                                      default_types=types
-                                      )
+
+TemplateValidator = validators.create(meta_schema=Draft4Validator.META_SCHEMA,
+                                      validators=Draft4Validator.VALIDATORS)
+format_checker = FormatChecker()
+
+TYPE_STRINGS = ('int', 'int8', 'int16', 'int32', 'int64',
+                'float', 'float8', 'float16', 'float32', 'float64', 'double',
+                'S1', 'c', 'i1', 'b', 'B', 'u1', 'i2', 'h', 's', 'u2', 'i4', 'i', 'l', 'u4',
+                'i8', 'u8', 'f4', 'f', 'f8', 'd'
+                )
+
+
+@format_checker.checks('datatype')
+def is_datatype(value):
+    """Return whether the given value is a valid data type specification for a NetCDF variable"""
+    if isinstance(value, basestring):
+        if value in TYPE_STRINGS:
+            return True
+        else:
+            return False
+    if isinstance(value, np.dtype):
+        return True
+    if isinstance(value, type):
+        if issubclass(value, np.number):
```

Comment:
Already a bool function, so this can just be

```python
return issubclass(value, np.number)
```

Also, this suggests that sometimes it is receiving an *instance* and sometimes it is receiving a *class* reference? What's the rationale behind that?

Reply:
This is validating values that can be passed to the `datatype` arg of netCDF [`createVariable`](http://unidata.github.io/netcdf4-python/#netCDF4.Dataset.createVariable) method. The docs say:
> The datatype can be a numpy datatype object, or a string that describes a numpy dtype object (like the dtype.str attribute of a numpy array). Supported specifiers include: 'S1' or 'c' (NC_CHAR), 'i1' or 'b' or 'B' (NC_BYTE), 'u1' (NC_UBYTE), 'i2' or 'h' or 's' (NC_SHORT), 'u2' (NC_USHORT), 'i4' or 'i' or 'l' (NC_INT), 'u4' (NC_UINT), 'i8' (NC_INT64), 'u8' (NC_UINT64), 'f4' or 'f' (NC_FLOAT), 'f8' or 'd' (NC_DOUBLE)

So that means _instances_ of `numpy.dtype`, or the strings listed. From experimenting, I know it can also accept more informative strings like 'float32', 'int64', etc..., as well as _classes_ such as `numpy.float32`, which are subclasses of `numpy.number`.

So to maintain this flexibility, we need to allow for all these possibilities.

Reply:
Alternatively, I could just not validate to this level, and allow the netCDF4 library to complain about anything it can't make sense of. That would simplify the verification code here, but in order to give more helpful error messages, we'd have to catch these exceptions and add more info, so there would be more overhead in the `DatasetTemplate` methods that write the file.